wordpress-mobile/gutenberg-mobile
# Adding retry feature to mobile apiFetch handler

This PR is a simple native-only solution to retry networking calls when they fail with unknown errors. Fixing point 2 of #1593 .

- WPiOS side PR: https://github.com/wordpress-mobile/WordPress-iOS/pull/13119
- WPAndroid side PR: https://github.com/wordpress-mobile/WordPress-Android/pull/11001
- FluxC (Android) PR: https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/1459 (merged)

The idea is to catch the error before it returns to the caller, and retrying it if the error is unknown.

There are many kind of errors (both 4xx and 5xx) and we probably don't want to retry, so I decided to go simplistic and focus on caching the most common errors on mobile, that is no internet connection or time out for bad connectivity.

This approach has the following problems (with our only networking call so far [image object] as an example). Consider this scenario:

- An image block requested the image object and a network call was triggered.
- In the case the network call failed because of network issues, we will retry it.
- In the mean time, the Image block was deleted from the post.
- ⚠️ apiFetch has no way to know that this request is not needed anymore, and will continue retrying forever (until the request succeeds or Gutenberg is closed).
- To solve this, I decided to add a maximum amount of retries (20 for now).
- ⚠️ After all the retries attempt fails, there's no way to get the image object for this particular image ever again in the current Gutenberg session.

I think this is still better than never retrying anything, but it's also not the ideal.

The values I have chosen out of trial and error:
- Retry each 2 seconds.
- 20 retries max.

Not big feelings about those, but it seemed good enough to not send requests too often, and keep retrying for a decent amount of time.

---

## Android side work left to do

UPDATE: Android side has been added

~For this to work, the error returned from the parent apps needs to include the HTTP Status code. WPiOS needed a small change allow that.~

~I noticed that WPAndroid doesn't return status code either, so we will need a fix for this on there too. @mchowning could you take a look at it?~

To test:
- For WPiOS: Checkout https://github.com/wordpress-mobile/WordPress-iOS/pull/13119
- Add an image block.
- Choose to add an image from WP Media Gallery.
- Turn off internet connection.
- Choose the image.
- Check that image doesn't load.
- Enable internet connection.
- Check that (eventually) the image will load.
- Check that Image Sizes settings work.

PR submission checklist:
